### PR TITLE
Compound library - types, instances and pagination

### PR DESCRIPTION
**Pull request by iamvigneshwars**
I made changes to the compound library, like we discussed I've created two tables compound_types and compound_instances with one to many relationship between them.

### CODE REVIEW

File: backend/Cargo.toml [hunk 1]
```diff
@@ -45,4 +46,4 @@ tokio = { version = "1.35.1", features = ["macros", "rt-multi-thread"] }
 tracing = { version = "0.1.40" }
 tracing-subscriber = { version = "0.3.18", features = ["env-filter"] }
 url = { version = "2.5.0" }
-uuid = { version = "1.7.0", features = ["v4"] }
+uuid = { version = "1.7.0", features = ["v4", "v7"] }
```
Comment (garryod) [inline backend/Cargo.toml:49]: Would like to just have one of `v4` & `v7`, can we merge an update to move current ones to `v7` before this?

File: .gitignore [hunk 2]
```diff
@@ -3,3 +3,5 @@
 
 # Developer Tooling
 .vscode
+
+router
```
Comment (garryod) [inline .gitignore:7]: - A comment to say what this relates to would be useful
- Needs newline at EOF

File: .devcontainer/Dockerfile.backend [hunk 3]
```diff
@@ -7,3 +7,6 @@ RUN apt-get update \
     sqlite3 pre-commit \
     libopencv-dev clang libclang-dev \
     && rm -rf /var/lib/apt/lists/*
+
+RUN curl -sSL https://rover.apollo.dev/nix/latest | sh \
+    && curl -sSL https://router.apollo.dev/download/nix/latest | sh
```
Comment (garryod) [inline .devcontainer/Dockerfile.backend:12]: Needs newline at EOF

File: backend/compound_library/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "compound_library"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
```
Comment (garryod) [inline backend/compound_library/Cargo.toml:7]:
```suggestion
```

File: backend/compound_library/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+dotenvy ={ workspace = true }
+clap ={ workspace = true }
+opa_client ={ path = "../opa_client", features = ["graphql"]}
+url ={ workspace = true }
+tracing = { workspace = true }
+tracing-subscriber = { workspace = true }
+uuid = { workspace = true , features = ["v7"]}
+the_paginator = { version = "0.1.0", path = "../the_paginator", features = [
+    "async-graphql",
+] }
```
Comment (garryod) [inline backend/compound_library/Cargo.toml:24]: Formatting is a bit weird here, any chance you could sort it alphabetically and format it with Tapio / VSCode Even better TOML (`tamasfe.even-better-toml`)

File: backend/compound_library/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+sea-orm-migration ={ workspace = true }
+async-graphql ={ workspace = true }
+graphql_endpoints = { path = "../graphql_endpoints" }
+dotenvy ={ workspace = true }
+clap ={ workspace = true }
+opa_client ={ path = "../opa_client", features = ["graphql"]}
+url ={ workspace = true }
+tracing = { workspace = true }
+tracing-subscriber = { workspace = true }
+uuid = { workspace = true , features = ["v7"]}
```
Comment (garryod) [inline backend/compound_library/Cargo.toml:21]: Shouldn't need this feature if it's defined in the workspace

File: backend/compound_library/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "compound_library"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+sea-orm ={ workspace = true, features = ["sqlx-postgres"]}
```
Comment (garryod) [inline backend/compound_library/Cargo.toml:9]: Shouldn't need this feature if it's defined in the workspace

File: backend/compound_library/src/entities/compound_instances.rs [hunk 8]
```diff
@@ -0,0 +1,38 @@
+//src/entities/compound_instances.rs
+
```
Comment (garryod) [inline backend/compound_library/src/entities/compound_instances.rs:2]:
```suggestion
```

File: backend/compound_library/src/entities/compound_types.rs [hunk 9]
```diff
@@ -0,0 +1,34 @@
+//src/entities/compound_types.rs
+
```
Comment (garryod) [inline backend/compound_library/src/entities/compound_types.rs:2]:
```suggestion
```

File: backend/compound_library/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,113 @@
+// main.rs
+
```
Comment (garryod) [inline backend/compound_library/src/main.rs:2]:
```suggestion
```

File: backend/compound_library/src/migrator.rs [hunk 11]
```diff
@@ -0,0 +1,36 @@
+// src/migrator.rs
+
```
Comment (garryod) [inline backend/compound_library/src/migrator.rs:2]:
```suggestion
```

File: backend/compound_library/src/resolvers/compound_instances_res.rs [hunk 12]
```diff
@@ -0,0 +1,80 @@
+// src/resolvers/compound_instances_res.rs
+
```
Comment (garryod) [inline backend/compound_library/src/resolvers/compound_instances_res.rs:2]:
```suggestion
```

File: backend/compound_library/src/resolvers/compound_types_res.rs [hunk 13]
```diff
@@ -0,0 +1,81 @@
+// src/resolvers/compound_types_res.rs
+
```
Comment (garryod) [inline backend/compound_library/src/resolvers/compound_types_res.rs:2]:
```suggestion
```

File: backend/supergraph.graphql
Comment (garryod) [inline backend/supergraph.graphql:1]: This file appears to be generated, if true, can we avoid committing it?

File: policies/compound_library.rego [hunk 14]
```diff
@@ -0,0 +1,23 @@
+package xchemlab.compound_library
+
+import data.xchemlab
+import future.keywords.if
```
Comment (garryod) [inline policies/compound_library.rego:4]: Importing `rego.v1` is preferred to `futures.x` as it ops into all of the breaking changes that will come with OPA v1

```suggestion
import rego.v1
```

File: policies/compound_library.rego [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    }
+}
+
+write_compound = response if {
+    xchemlab.valid_token
+    response := {
+        "allowed" : true, 
+        "subject" : xchemlab.subject,
+    }
+}
```
Comment (garryod) [inline policies/compound_library.rego:23]: Need a newline at EOF

File: backend/router.yaml
Comment (garryod) [inline backend/router.yaml:1]: Can we move the supergraph stuff to another PR?

File: backend/supergraph-config.yaml
Comment (garryod) [inline backend/supergraph-config.yaml:1]: Can we move the supergraph stuff to another PR?

**Review by garryod — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 2 above)
Comment (iamvigneshwars) [reply .gitignore:7]: I've removed router for now, we can add this in a separate PR for supergraphs

File: .devcontainer/Dockerfile.backend (hunk 3 above)
Comment (iamvigneshwars) [reply .devcontainer/Dockerfile.backend:12]: I've remove this, we can add it with the supergraph PR

File: policies/compound_library.rego (hunk 14 above)
Comment (iamvigneshwars) [reply policies/compound_library.rego:4]: I guess future.keywords.if is used for other services as well. But I've changed for this service

File: backend/router.yaml
Comment (iamvigneshwars) [reply backend/router.yaml:1]: yes, we can more it to another PR, I've removed all the supergraph stuffs here.

File: backend/compound_library/src/graphql.rs [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+    compound_instances_res::{CompoundInstanceMutation, CompoundInstanceQuery},
+    compound_types_res::{CompoundMutation, CompoundQuery},
+};
+use async_graphql::{EmptySubscription, MergedObject, Schema, SchemaBuilder};
+
+#[derive(Debug, Clone, MergedObject, Default)]
+pub struct Query(CompoundQuery, CompoundInstanceQuery);
+
+#[derive(Debug, Clone, MergedObject, Default)]
+pub struct Mutation(CompoundMutation, CompoundInstanceMutation);
```
Comment (garryod) [inline backend/compound_library/src/graphql.rs:11]: Imo, `RootMutation` is more descriptive

File: backend/compound_library/src/resolvers/compound_instances_res.rs [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    async fn add_compound_instance(
+        &self,
+        ctx: &Context<'_>,
+        plate_id: Uuid,
+        well_num: i16,
+        compound_type: Uuid,
+    ) -> async_graphql::Result<compound_instances::Model> {
+        subject_authorization!("xchemlab.compound_library.write_compound", ctx).await?;
+        let db = ctx.data::<DatabaseConnection>()?;
+        let compound_instance = compound_instances::ActiveModel {
```
Comment (garryod) [inline backend/compound_library/src/resolvers/compound_instances_res.rs:66]: We should probably also store who added this and when

File: backend/compound_library/src/resolvers/compound_instances_res.rs [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+#[derive(Debug, Clone, Default)]
+pub struct CompoundInstanceMutation;
+
+#[ComplexObject]
+impl compound_instances::Model {
+    async fn compound_types(
+        &self,
+        ctx: &Context<'_>,
+    ) -> async_graphql::Result<Vec<compound_types::Model>> {
```
Comment (garryod) [inline backend/compound_library/src/resolvers/compound_instances_res.rs:21]: Doesn't this always return one instance?

File: backend/compound_library/src/resolvers/compound_instances_res.rs [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+    }
+}
+
+#[Object]
+impl CompoundInstanceQuery {
+    async fn compound_instances(
+        &self,
+        ctx: &Context<'_>,
+        cursor: CursorInput,
+    ) -> async_graphql::Result<ModelConnection<compound_instances::Model>> {
```
Comment (garryod) [inline backend/compound_library/src/resolvers/compound_instances_res.rs:34]: Do we want to be able to filter this?

File: backend/compound_library/src/resolvers/compound_types_res.rs [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+#[Object]
+impl CompoundMutation {
+    async fn add_compound(
+        &self,
+        ctx: &Context<'_>,
+        name: String,
+        smiles: String,
+    ) -> async_graphql::Result<compound_types::Model> {
```
Comment (garryod) [inline backend/compound_library/src/resolvers/compound_types_res.rs:65]: Do we want to record who added this and when?

File: backend/compound_library/src/resolvers/compound_types_res.rs [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+            .try_into_query_cursor::<compound_types::Entity>()?
+            .all(db)
+            .await?
+            .try_into_connection()?)
+    }
+
+    async fn compound(
+        &self,
+        ctx: &Context<'_>,
+        id: Uuid,
```
Comment (garryod) [inline backend/compound_library/src/resolvers/compound_types_res.rs:50]: Would it make sense to select by name instead of ID?

**Review by garryod — CHANGES_REQUESTED** [review]

File: backend/pin_packing/src/graphql.rs [hunk 22]
```diff
@@ -9,18 +9,19 @@ use crate::resolvers::{
...
     )
+    .enable_federation()
 }
 
-pub type RootSchema = Schema<RootQuery, RootMutation, EmptySubscription>;
+pub type RootSchema = Schema<Query, Mutation, EmptySubscription>;
 
 #[derive(Debug, Clone, MergedObject, Default)]
-pub struct RootQuery(
+pub struct Query(
```
[the remark this answers is not in the document]
Comment (iamvigneshwars) [reply backend/pin_packing/src/graphql.rs]: It was for the federation stuff, I've changed it back.

File: policies/compound_library.rego (hunk 14 above)
Comment (garryod) [reply policies/compound_library.rego]: Ah, probably want to change it for all of them, but I can do that at a later date

File: backend/compound_library/src/resolvers/compound_types_res.rs (hunk 20 above)
Comment (iamvigneshwars) [reply backend/compound_library/src/resolvers/compound_types_res.rs:65]: Record in the database?

Comment (garryod) [reply backend/compound_library/src/resolvers/compound_types_res.rs:65]: Yeah

File: backend/compound_library/src/graphql.rs [hunk 23]
```diff
@@ -0,0 +1,17 @@
+use crate::resolvers::{
+    compound_instances_res::{CompoundInstanceMutation, CompoundInstanceQuery},
+    compound_types_res::{CompoundMutation, CompoundQuery},
+};
+use async_graphql::{EmptySubscription, MergedObject, Schema, SchemaBuilder};
+
+#[derive(Debug, Clone, MergedObject, Default)]
+pub struct Query(CompoundQuery, CompoundInstanceQuery);
```
[the remark this answers is not in the document]
Comment (iamvigneshwars) [reply backend/compound_library/src/graphql.rs:8]: We have to use Query and Mutation when we work with federation stuff.

File: backend/compound_library/src/graphql.rs [hunk 24]
```diff
@@ -0,0 +1,23 @@
+use crate::resolvers::{
+    compound_types_res::{CompoundMutation, CompoundQuery},
+    compound_instances_res::{CompoundInstanceMutation, CompoundInstanceQuery}
+};
+use async_graphql::{EmptySubscription, MergedObject, Schema, SchemaBuilder};
+
+#[derive(Debug, Clone, MergedObject, Default)]
+pub struct Query(
```
[the remark this answers is not in the document]
Comment (iamvigneshwars) [reply backend/compound_library/src/graphql.rs]: Apollo federation only works if its Query

File: backend/compound_library/src/graphql.rs [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+    CompoundQuery,
+    CompoundInstanceQuery,
+);
+
+#[derive(Debug, Clone, MergedObject, Default)]
+pub struct Mutation(
```
[the remark this answers is not in the document]
Comment (iamvigneshwars) [reply backend/compound_library/src/graphql.rs]: Apollo federation only works if its Mutation

File: backend/compound_library/src/resolvers/compound_instances_res.rs (hunk 19 above)
Comment (iamvigneshwars) [reply backend/compound_library/src/resolvers/compound_instances_res.rs:34]: we can already filter the pages right? or did you mean something else?

File: backend/compound_library/src/resolvers/mod.rs [hunk 26]
```diff
@@ -0,0 +1,2 @@
+pub mod compound_types_res;
```
[the remark this answers is not in the document]
Comment (iamvigneshwars) [reply backend/compound_library/src/resolvers/mod.rs]: Yeah, we need to access it outside the resolvers module.

File: backend/compound_library/src/graphql.rs (hunk 16 above)
Comment (iamvigneshwars) [reply backend/compound_library/src/graphql.rs:11]: Needed for Apollo federation.

File: backend/compound_library/src/entities/compound_types.rs [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel, SimpleObject)]
+#[sea_orm(table_name = "compound_types")]
+#[graphql(name = "compound_types", complex)]
+pub struct Model {
+    #[sea_orm(primary_key, auto_increment = false)]
+    pub id: Uuid,
+    #[sea_orm(column_type = "Text", unique)]
+    pub name: String,
+    pub smiles: String,
```
[the remark this answers is not in the document]
Comment (iamvigneshwars) [reply backend/compound_library/src/entities/compound_types.rs:19]: I guess its much straightforward just to use varchar

File: backend/compound_library/Cargo.toml (hunk 7 above)
Comment (iamvigneshwars) [reply backend/compound_library/Cargo.toml]: Feature is not defined in the workspace

File: backend/compound_library/src/resolvers/compound_instances_res.rs (hunk 19 above)
Comment (garryod) [reply backend/compound_library/src/resolvers/compound_instances_res.rs:33]: Yeah, like filtering by plate, or well number. Probably not needed now, but worth considering how we'd do it

File: backend/compound_library/src/resolvers/mod.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (garryod) [reply backend/compound_library/src/resolvers/mod.rs]: Ah, for the `graphql` stuff... Yeah, should probably make `graphql` the directory and merge it with the resolvers... Not sure what I was thinking there

File: backend/compound_library/src/entities/compound_types.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (garryod) [reply backend/compound_library/src/entities/compound_types.rs:16]: It's the easiest way of doing things, but it allows you to store invalid sequences, perhaps we should store it as that with some validation?

File: backend/compound_library/src/graphql.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (garryod) [reply backend/compound_library/src/graphql.rs]: Ah, thought you said it was the other way around earlier, oops

File: backend/compound_library/src/graphql.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (garryod) [reply backend/compound_library/src/graphql.rs:8]: Ah, thought you said it was the other way around earlier, oops

File: backend/compound_library/src/graphql.rs (hunk 16 above)
Comment (garryod) [reply backend/compound_library/src/graphql.rs:11]: Ah, thought you said it was the other way around earlier, oops

File: backend/compound_library/Cargo.toml (hunk 7 above)
Comment (garryod) [reply backend/compound_library/Cargo.toml]: Ah yeah, forgot we use `sqlx-sqlite` for one of the crates :facepalm:

File: backend/compound_library/src/main.rs [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+#[derive(Debug, Parser)]
+struct SchemaArgs {
+    #[arg(short, long)]
+    path: Option<PathBuf>,
+}
+
+async fn setup_database() -> Result<DatabaseConnection, TransactionError<DbErr>> {
+    let db_url =
+        ConnectOptions::new("postgres://postgres:password@postgres/compound_library".to_string());
```
[the remark this answers is not in the document]
Comment (iamvigneshwars) [reply backend/compound_library/src/main.rs:46]: For now, I'm just using the full URL <code>COMPOUND_LIBRARY_DATABASE_URL</code>  we can think about this later, I guess it depends on how the services are gonna be deployed.

[the remark this answers is not in the document]
Comment (garryod) [reply backend/compound_library/src/main.rs:46]: They'll each be in their own K8s pod, so a unique environment variable isn't required there. I guess another options is that they could have a default `database` if the URL doesn't contain one (a path)

[the remark this answers is not in the document]
Comment (iamvigneshwars) [reply backend/compound_library/src/main.rs]: Yeah, that's a good idea, I'll add a default unique path.

File: backend/compound_library/src/graphql/schema.rs
Comment (garryod) [inline backend/compound_library/src/graphql/schema.rs:1]: Should this just go in `mod`?

File: backend/pin_packing/src/graphql.rs [hunk 29]
```diff
@@ -15,6 +15,7 @@ pub fn root_schema_builder() -> SchemaBuilder<RootQuery, RootMutation, EmptySubs
         RootMutation::default(),
         EmptySubscription,
     )
+    .enable_federation()
```
Comment (garryod) [inline backend/pin_packing/src/graphql.rs:18]: Do we need this yet?

File: backend/compound_library/src/graphql/schema.rs [hunk 30]
```diff
@@ -0,0 +1,18 @@
...
+#[derive(Debug, Clone, MergedObject, Default)]
+pub struct Query(CompoundQuery, CompoundInstanceQuery);
+
+#[derive(Debug, Clone, MergedObject, Default)]
+pub struct Mutation(CompoundMutation, CompoundInstanceMutation);
+
+pub type RootSchema = Schema<Query, Mutation, EmptySubscription>;
+
+pub fn root_schema_builder() -> SchemaBuilder<Query, Mutation, EmptySubscription> {
+    Schema::build(Query::default(), Mutation::default(), EmptySubscription).enable_federation()
```
Comment (garryod) [inline backend/compound_library/src/graphql/schema.rs:17]: Do we need `enable_federation()` yet?

File: backend/compound_library/src/graphql/compound_types_res.rs [hunk 31]
```diff
@@ -0,0 +1,123 @@
...
+use async_graphql::{ComplexObject, Context, Object};
+use chrono::Utc;
+use opa_client::subject_authorization;
+use purr::{
+    graph::Builder,
+    read::{read, Trace},
+};
+use sea_orm::{ActiveValue, ColumnTrait, DatabaseConnection, EntityTrait, ModelTrait, QueryFilter};
+use the_paginator::graphql::{CursorInput, ModelConnection};
+// use uuid::Uuid;
```
Comment (garryod) [inline backend/compound_library/src/graphql/compound_types_res.rs:11]:
```suggestion
```

**Review by garryod — CHANGES_REQUESTED** [review]
Looking pretty close to complete, any chance you could set the following in `main.rs` and fix the warnings?

```rust
#![forbid(unsafe_code)]
#![warn(missing_docs)]
#![warn(clippy::missing_docs_in_private_items)]
#![doc=include_str!("../README.md")]
```

File: backend/compound_library/src/main.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (garryod) [reply backend/compound_library/src/main.rs]: Sweet, then we can just use `DATABASE_URL`

File: backend/pin_packing/src/graphql.rs (hunk 29 above)
Comment (iamvigneshwars) [reply backend/pin_packing/src/graphql.rs:18]: Nope, will remove it!

File: backend/compound_library/src/graphql/compound_types_res.rs [hunk 32]
```diff
@@ -0,0 +1,108 @@
...
+    }
+}
+
+impl CustomValidator<String> for SmilesValidator {
+    fn check(&self, smiles: &String) -> Result<(), InputValueError<String>> {
+        if self.valid_smiles(smiles) {
+            Ok(())
+        } else {
+            Err(InputValueError::custom("Invalid SMILES format"))
+        }
```
Comment (garryod) [inline backend/compound_library/src/graphql/compound_types_res.rs:83]: I believe this is considered more idiomatic

```suggestion
        self.valid_smiles(smiles).then_some(()).ok_or(InputValueError::custom("Invalid SMILES format"))
```

File: backend/compound_library/src/graphql/compound_types_res.rs [hunk 33]
```diff
@@ -0,0 +1,108 @@
...
+
+#[derive(Debug, Clone, Default)]
+pub struct CompoundMutation;
+
+#[ComplexObject]
+impl compound_types::Model {
+    async fn compound_instances(
+        &self,
+        ctx: &Context<'_>,
+    ) -> async_graphql::Result<Vec<compound_instances::Model>> {
```
Comment (garryod) [inline backend/compound_library/src/graphql/compound_types_res.rs:23]: `instances` might be a more appropriate name since we're already on a `CompoundType` node

File: backend/compound_library/src/graphql/compound_instances_res.rs [hunk 34]
```diff
@@ -0,0 +1,80 @@
...
+
+#[derive(Debug, Clone, Default)]
+pub struct CompoundInstanceMutation;
+
+#[ComplexObject]
+impl compound_instances::Model {
+    async fn compound_types(
+        &self,
+        ctx: &Context<'_>,
+    ) -> async_graphql::Result<Option<compound_types::Model>> {
```
Comment (garryod) [inline backend/compound_library/src/graphql/compound_instances_res.rs:20]: `type` may be a more appropriate name as we're already on a `CompoundInstance` node

File: backend/compound_library/src/entities/compound_types.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (iamvigneshwars) [reply backend/compound_library/src/entities/compound_types.rs:19]: I have validated using the graphql custom validation. I tried using <code>ActiveModelBehaviour</code> initially but it did not trigger the <code>before_save()</code> or <code>after_save</code> for some reason.

[the remark this answers is not in the document]
Comment (garryod) [reply backend/compound_library/src/entities/compound_types.rs:19]: Hmm, that's rather odd, are there any issues in at the moment regarding it?

[the remark this answers is not in the document]
Comment (iamvigneshwars) [reply backend/compound_library/src/entities/compound_types.rs:19]: No there are no issues ATM, it works fine with the graphql  custom validation

[the remark this answers is not in the document]
Comment (garryod) [reply backend/compound_library/src/entities/compound_types.rs:19]: Strange, might pop over and take a look. The `async-graphql` validator looks good though

File: backend/compound_library/src/entities/compound_instances.rs [hunk 35]
```diff
@@ -0,0 +1,50 @@
...
+#[graphql(name = "compund_instances", complex)]
+pub struct Model {
+    /// A `Uuid` uniquely identifies the plate
+    #[sea_orm(primary_key, auto_increment = false)]
+    pub plate_id: Uuid,
+    /// An `i16` that serves as the other part of the primary key, uniquely identifying
+    /// the well within the plate where the compound instance is located.
+    #[sea_orm(primary_key, auto_increment = false)]
+    pub well_number: i16,
+    /// A `String` that describes the type of the compound in the well.
```
Comment (garryod) [inline backend/compound_library/src/entities/compound_instances.rs:22]:
```suggestion
    /// The name of the type of the compound in the well.
```

File: backend/compound_library/src/entities/compound_instances.rs [hunk 36]
```diff
@@ -0,0 +1,50 @@
...
+    EnumIter, PrimaryKeyTrait, Related, RelationDef, RelationTrait,
+};
+use uuid::Uuid;
+
+/// Represents a compound instance within the database.
+#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel, SimpleObject)]
+#[sea_orm(table_name = "compound_instances")]
+#[graphql(name = "compund_instances", complex)]
+pub struct Model {
+    /// A `Uuid` uniquely identifies the plate
```
Comment (garryod) [inline backend/compound_library/src/entities/compound_instances.rs:15]: ID of the plate on which the compound instance is located.

File: backend/compound_library/src/entities/compound_instances.rs [hunk 37]
```diff
@@ -0,0 +1,50 @@
...
+    #[sea_orm(primary_key, auto_increment = false)]
+    pub plate_id: Uuid,
+    /// An `i16` that serves as the other part of the primary key, uniquely identifying
+    /// the well within the plate where the compound instance is located.
```
Comment (garryod) [inline backend/compound_library/src/entities/compound_instances.rs:19]:
```suggestion
    The well on the plate which the compound instance is located.
```

File: backend/compound_library/src/entities/compound_instances.rs [hunk 38]
```diff
@@ -0,0 +1,50 @@
...
+    #[sea_orm(primary_key, auto_increment = false)]
+    pub well_number: i16,
+    /// A `String` that describes the type of the compound in the well.
+    pub compound_type: String,
+    /// A `String` representing the identifier of the operator who
+    /// processed or handled the compound instance.
```
Comment (garryod) [inline backend/compound_library/src/entities/compound_instances.rs:25]:
```suggestion
    /// The identifier of the operator which added this entry.
```

File: backend/compound_library/src/entities/compound_instances.rs [hunk 39]
```diff
@@ -0,0 +1,50 @@
...
+    pub operator_id: String,
+    /// A `DateTime<Utc>` indicating the date and time when the compound
+    /// instance was processed or recorded.
```
Comment (garryod) [inline backend/compound_library/src/entities/compound_instances.rs:28]:
```suggestion
    /// The date and time when the compound instance was added.
```

**Review by garryod — COMMENTED** [review]
Looking good, only one comment which is that when it comes to struct fields we can omit anything about the type of the field in the docstring as this is captured by the type itself, see below for a few examples...

**Review by garryod — APPROVED** [review]
